rpc/connection_log.go

```diff
@@ -0,0 +1,123 @@
+// Copyright 2018 Keybase, Inc. All rights reserved. Use of
+// this source code is governed by the included BSD license.
+
+package rpc
+
+import (
+	"encoding/hex"
+	"fmt"
+	"math/rand"
+	"runtime"
+)
+
+type logField struct {
+	key   string
+	value interface{}
+}
+
+// Format implements the fmt.Formatter interface, to make the structured
+// logField compatible with format-based non-structured loggers.
+func (f logField) Format(s fmt.State, verb rune) {
+	fmt.Fprintf(s, "%"+string(verb), f.value)
+}
+
+// connectionLog defines an interface used by connection.go for logging. An
+// implementation typically only uses either of `msg` and `format` field,
```

Comment:
It's pretty confusing to have two different string fields.  Can we just have the `format` field, and pass the fully formatted string to logrus?  I realize the field values will be duplicated in that case, but it seems less confusing to me.

Reply:
> pass the fully formatted string to logrus

If I understand correctly, this makes the same kind of messages have different strings, which don't get properly indexed by kibana, making it difficult to find.

Reply:
+1 to strib's comment. We shouldn't create a new API just to support structuring. There's too much risk to changing this too. We need to implement the same API for structured logs as we do today.

Reply:
@songgao: I was thinking that if you need an indexible field, you could just add a new log field with the message type.

Reply:
> We need to implement the same API for structured logs as we do today.

This is in `rpc` though, so if we just use the existing structured log API, existing log format would change. If you guys think it's OK to change logging format in clients, I'm happy to use `msg` instead of `format` everywhere.

> if you need an indexible field, you could just add a new log field with the message type.

That's what the added `msg` is right?

Reply:
One possible answer is that if the API is:
```go
type connectionLog interface {
    Log(msg string, params ...interface{})
}
```
Then an unstructured logger can treat `msg` as a format string (or `FormattedLogger` as the case may be). A structured logger can treat `msg` as a base message, and hopefully the log fields as actual content.

Alternatively, maybe the approach here is wrong altogether, and instead of trying to shoehorn both loggers into one interface, we should just have a:
```go
type logFields map[string]interface{}

type connectionLog interface {
    Log(msg string, params ...interface{})
    LogWithFields(msg string, fields logFields)
    Debug(msg string, params ...interface{})
    DebugWithFields(msg string, fields logFields)
    ...
}
```
In that case, the old API would remain, and all current implementing receivers would continue to use that, except where we want to support the new API. That way, an unstructured logger can choose a default way to represent `LogWithFields`, and a structured logger can choose a default way to represent `Log`.

Reply:
Hmm that seems to require each place where logging is needed to call two versions of logging, e.g. one on `Debug` and one on `DebugWithFields` since the caller doesn't know which logger we are using underlying.

Reply:
Nope, it doesn't require anything to change. The caller decides which to use, and the implementing logger has sane defaults for the opposite situation.

Reply:
> Then an unstructured logger can treat msg as a format string (or FormattedLogger as the case may be). A structured logger can treat msg as a base message, and hopefully the log fields as actual content.

Yes this would work, and was actually the initial direction I was going. But it inevitably brings changes to the actual logging format in clients as I mentioned above. If that's an acceptable compromise, I'm happy to change!

Reply:
> The caller decides which to use, and the implementing logger has sane defaults for the opposite situation.

The caller is `connection.go` in this package. The structured logger is in client, and unstructured in servers. We need the same messages in both cases.

Reply:
I don't see why anything has to change...

Reply:
Oh you were talking about the first alternative. Yeah I think I convinced myself that's the wrong way to go.

Reply:
My comments above about needing to call both `Debug` and `DebugWithFields` are about the second alternative. The part about actual logging content appearing differently is about the first alternative.

Reply:
hmm so which way are we going? @jzila @strib 

Also just to make it clear, this logger is in `connection.go` only, and doesn't impact other parts of the `rpc` package at all.

Reply:
I have to think about it some more. Please hold off merging.

Reply:
Sure; wasn't about to merge yet. No rush!

Reply:
OK I think I'm convinced by @strib's approach.

Your messages will look like this:
```go
const logMessagePrimaryKeyName = "key"

c.log.Debug("Connection: %s; status: %d",
    logField{key: logMessagePrimaryKeyName, value: "waitForConnection"},
    logField{key: "disconnectStatus", value: disconnectStatus})
```

Reply:
Your structured logger can choose to ignore the format string if it wants to, or it can include it.

It's a bit restrictive because your log fields must happen in the same order as the format string, but I think that might actually be a good thing.

Reply:
And for structured logging you can always include keys at the end that aren't in the format string. It's a little restrictive but also pretty flexible.

Reply:
(assuming you meant `const logMessagePrimaryKeyName = "msg")

So basically extracting `msg` in function argument from the format string I guess. Sounds good. The logging calls will be less-descriptive though. So for example, instead of:

```
         ct.log.Debug("dialing", "Dialing %s",
             logField{key: "remote-addr", value: addr})
```

We'll have

```
         ct.log.Debug("%s %s",
             logField{key: "msg", value: "Dialing"})
             logField{key: "remote-addr", value: addr})
```

Sounds good?

Reply:
You could then make a separate helper method if you like:
```go
type logMessagePrimaryKey string
func (l unstructuredLogger) DebugWithKey(key logMessagePrimaryKey, fmt string, params ...interface{}) {
    fmt.Printf("(%s) " + fmt, key, params...)
}

c.log.Debug("dialing", "ip %s", ip)
```

Reply:
Hmm I don't understand the added `DebugWithKey`. When would it be used? I thought we'd continue use `ct.log.Debug` everywhere.

Reply:
@strib: would that work for you? We'd have a separate type for the key, although log messages would still have 2 consecutive `string`s. I guess we could make them all the keys constants.

Reply:
@songgao: it's just a helper if you wanted the log messages to be more readable in code.

Reply:
That's fine.  I mainly just didn't like the two parameters and their names (`msg` and `format`).

Reply:
Hmm not sure if the help is gonna work. It would change both the interface for generating logging, *and* change the actual text that's being logged. Anyway, will implement this approach. THanks!